refactor(compiler): implement `ngDeclareInjectable()`

petebacondarwin opened the pull request:
This PR sits on top of #41231, so you only need to start your review from **refactor(compiler): fix typo**.

petebacondarwin committed:
refactor(compiler): implement `ngDeclareInjectable()`

This commit changes the partial compilation so that it outputs declarations
rather than definitions for injectables.

The JIT compiler and the linker are updated to be able to handle these
new declarations.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -371,6 +371,48 @@ export enum R3FactoryTarget {
+   */
+  useValue?: o.Expression;
+  /**
+   * An array of dependencies to support instantiating this injecttable via `useClass` or
```
```suggestion
   * An array of dependencies to support instantiating this injectable via `useClass` or
```

Looks like "injectable" is a hard word :-)

petebacondarwin replied:
Oh Man! It is nearly as hard as ~~unkown~~ ~~uknown~~ unknown.

JoostK commented on packages/compiler/src/render3/partial/injectable.ts:
```diff
@@ -0,0 +1,63 @@
+  definitionMap.set('type', meta.internalType);
+
+  // Only generate providedIn property if it has a non-null value
+  if ((meta.providedIn as o.LiteralExpr).value !== null) {
```
😱

Can we make `meta.providedIn` `null` instead? In fact, I believe this makes the comment on `R3DeclareInjectableMetadata` slightly inaccurate because it mentions that `providedIn` _can_ be set to `null`.

petebacondarwin replied:
I copied this code from here: https://github.com/angular/angular/blob/master/packages/compiler/src/injectable_compiler_2.ts#L115-L118

JoostK replied:
I see. That is a surprising bit of code then.

petebacondarwin replied:
As it turns out `definitionMap.set()` will ignore the property if the value is falsy, so this line is redundant anyway.

petebacondarwin replied:
Doh! But of course it is not falsy, it is an expression that represents `null` :-D

JoostK commented on packages/core/src/render3/jit/partial.ts:
```diff
@@ -42,6 +42,17 @@ export function ɵɵngDeclareFactory(decl: R3DeclareFactoryFacade): unknown {
+export function ɵɵngDeclareInjectable(decl: R3DeclareInjectableFacade): unknown {
+  const compiler = getCompilerFacade();
+  return compiler.compileInjectableDeclaration(
+      angularCoreEnv, `ng:///${decl.type.name}/ɵfac.js`, decl);
```
Shouldn't this be

```suggestion
      angularCoreEnv, `ng:///${decl.type.name}/ɵprov.js`, decl);
```

JoostK commented on packages/core/test/render3/jit/declare_injectable_spec.ts:
```diff
@@ -0,0 +1,141 @@
+ */
+
+import {InjectionToken, Injector, ɵcreateInjector, ɵɵdefineInjector, ɵɵInjectableDef, ɵɵngDeclareInjectable} from '@angular/core';
+import {forwardRef} from '@angular/core/src/di';
```
This one can be moved to the import above.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -371,6 +371,48 @@ export enum R3FactoryTarget {
+   * An array of dependencies to support instantiating this injecttable via `useClass` or
+   * `useFactory`.
+   */
+  userDeps?: R3DeclareDependencyMetadata[];
```
This is called `userDeps` in the internal metadata as there used to also be `ctorDeps`; probably from before factories were extracted. I'd rename this to just `deps` here as that aligns with the public name.

JoostK reviewed (commented):
This is a lot simpler than I was expecting it to be. Most stuff has already been done in the factory PR.

Can we also extend the compliance tests a bit, I haven't seen `useValue` nor `useExisting`.

petebacondarwin commented on packages/compiler/src/compiler_facade_interface.ts:
```diff
@@ -246,6 +248,16 @@ export enum ViewEncapsulation {
 export type ChangeDetectionStrategy = number;
 
+export interface R3DeclareInjectableFacade {
+  type: any;
+  providedIn?: any;
+  useClass?: any;
+  useFactory?: any;
+  useExisting?: any;
+  useValue?: any;
```
Yeah, I have been thinking about that. I think also that all the `type: any` in this file can become `type: unknown`, right?

petebacondarwin commented on packages/compiler/src/injectable_compiler_2.ts:
```diff
@@ -118,16 +114,18 @@ export function compileInjectable(meta: R3InjectableMetadata): InjectableDef {
 
+export function createInjectableType(meta: R3InjectableMetadata) {
+  return new o.ExpressionType(o.importExpr(
+      Identifiers.InjectableDef, [typeWithParameters(meta.type.type, meta.typeArgumentCount)]));
```
I didn't change this one in my previous PR precisely because I was worried it was already out in the public domain. But if you are confident that we never published it to npm in libraries then we should definitely change this.

petebacondarwin committed:
refactor(core): rename `ɵɵInjectableDef` interface to `ɵɵInjectableDeclaration`

The other similar interfaces were renamed in https://github.com/angular/angular/pull/41119,
but this one was left since it had existed before Ivy. It looks like the interface was
never actually exposed on npm so it is safe to rename this one too.

JoostK commented on packages/compiler/src/compile_metadata.ts:
```diff
@@ -36,6 +36,11 @@ export function identifierName(compileIdentifier: CompileIdentifierMetadata|null
+  if (ref['__forward_ref__']) {
+    // We do not want to try to stringify a `forwardRef()` function because that would cause the
+    // inner function to be evaluated too early, defeating the whole point of the `forwardRef`.
+    return '__forward_ref__';
```
This needs to be unique; you can probably repurpose the `_anonymousTypeIndex` counter

JoostK replied:
Oh, correction: it doesn't need to be unique! The emitter adds a counter itself. Brilliant.

petebacondarwin replied:
Yeah, I was going to mention this but you got there before me!

JoostK commented on packages/compiler/src/injectable_compiler_2.ts:
````diff
@@ -131,11 +131,16 @@ function delegateToFactory(type: o.WrappedNodeExpr<any>, internalType: o.Wrapped
+    // If types are different, we have to generate a wrapper function to ensure the internal type
+    // has been resolved. We also need to unwrap the `type` in case it is a `ForwardRef`.
+    // ```
+    // factory: function(t) { return core.resolveForwardRef(type).ɵfac(t); }
````
In the case I worked on earlier this was only relevant for JIT, so I then changed only the JIT compiled output. Can we do the same here? That saves some generated code for regular AOT compiles.

petebacondarwin replied:
Yeah - I plan to try. Currently I am just hacking commits to get things working. (WIP!)

petebacondarwin replied:
Looking at this https://github.com/angular/angular/commit/e23fd1f38205410e0ecb601ec73847cea2dea2a8 I think that what you did was added the `forwardRef`s to the partial-compiled output, and then in the linker you stripped them back out, right?

JoostK commented on packages/compiler/src/identifiers.ts:
```diff
@@ -126,6 +126,7 @@ export class Identifiers {
   };
   static createComponentFactory: o.ExternalReference = {name: 'ɵccf', moduleName: CORE};
   static setClassMetadata: o.ExternalReference = {name: 'ɵsetClassMetadata', moduleName: CORE};
+  static resolveForwardRef: o.ExternalReference = {name: 'resolveForwardRef', moduleName: CORE};
```
This already exists in Ivy's `Identifiers` btw.

petebacondarwin replied:
Ah cool.

petebacondarwin replied:
Actually, `forwardRef` and `resolvedForwardRef` are not specific to R3, so I think they should be in this file not the `r3_identifier.ts` file.

petebacondarwin replied:
But... I guess that only the Ivy compilation pipeline uses them so...

petebacondarwin replied:
But then we have this weird situation where `ɵɵdefineInjectable` is used outside of R3 but really fits with the R3 identifiers, so I am going to move that one and just live with it in the `packages/compiler/src/injectable_compiler.ts` file.

petebacondarwin committed:
refactor(compiler): fix typo

There was  spelling mistake in a comment.

JoostK commented on goldens/public-api/core/core.d.ts:
```diff
@@ -683,14 +683,6 @@ export declare function ɵɵdefineInjectable<T>(opts: {
 export declare function ɵɵinject<T>(token: Type<T> | AbstractType<T> | InjectionToken<T>, flags?: InjectFlags): T | null;
 
-/** @codeGenApi */
-export declare interface ɵɵInjectableDef<T> {
```
I also can't tell you why the rename causes it to be dropped from the goldens.

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_injectable_linker_1.ts:
```diff
@@ -0,0 +1,71 @@
+  const meta: R3InjectableMetadata = {
+    name: typeName,
+    type: wrapReference(typeExpr.getOpaque()),
+    internalType: metaObj.getOpaque('type'),
```
```suggestion
    internalType: typeExpr.getOpaque(),
```

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/util.ts:
```diff
@@ -29,3 +30,60 @@ export function parseEnum<TExpression, TEnum>(
+  const callee = expr.getCallee();
+  if (callee.getSymbolName() !== 'forwardRef') {
+    throw new FatalLinkerError(
+        callee.expression, 'Unsupported directive type, expected forwardRef or a type reference');
```
You may want to generalize this message a bit.

JoostK commented on packages/compiler/src/injectable_compiler_2.ts:
```diff
@@ -6,32 +6,54 @@
+   * If true, then the `expression` contains a reference to something that has not yet been
+   * defined.
+   *
+   * If so, then in certain circumstances (e.g. partial-compilation or JIT evaluation) the
+   * expression will need to be wrapped in a `forwardRef()` call.
```
I think this can more generally say that if `isForwardRef` is true, then `expression` must be emitted within a closure. For full compilations this happens to be the factory function, whereas partial compilation needs to reintroduce the `forwardRef`.

In the JIT case, the `expression` hasn't actually been unpacked to represent the forward-declared type itself as that would be a reference error. The JIT compiler should not actually repack this into a `forwardRef`, as that would result in a double-forward-ref (and it doesn't, so that's good). I believe this boolean is not actually used in JIT mode currently, as the `resolveForwardRef` call is always emitted for JIT as it has `resolveForwardRefs = true` passed into `compileInjectable`/`delegateToFactory`.

JoostK commented on packages/compiler/src/injectable_compiler_2.ts:
```diff
@@ -6,32 +6,54 @@
+}
+
+export function createR3ProviderExpression<T extends o.Expression>(
+    expression: T, isForwardRef = false): R3ProviderExpression<T> {
```
I think I would remove the default for `isForwardRef` here. The `R3ProviderExpression` type is meant to represent potentially unpacked forward references, so I'd expect the caller to be making a conscious decision about what its value should be.

JoostK commented on packages/compiler/src/injectable_compiler_2.ts:
```diff
@@ -118,25 +142,37 @@ export function compileInjectable(meta: R3InjectableMetadata): InjectableDef {
+function delegateToFactory(
+    type: o.WrappedNodeExpr<any>, internalType: o.WrappedNodeExpr<any>,
+    unwrapForwardRefs: boolean) {
+  const factoryContainer = unwrapForwardRefs ?
+      o.importExpr(Identifiers.resolveForwardRef).callFn([internalType]) :
+      internalType;
```
This is only used for the `type.node !== internalType.node` case; I think this code would become more readable (and the comment less awkwardly placed) if the ternary was converted into an if-else block.

petebacondarwin replied:
I'll clean this up a bit.

JoostK commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -60,18 +60,41 @@ export class CompilerFacadeImpl implements CompilerFacade {
+          useExisting: convertToProviderExpression(facade, USE_EXISTING),
+          deps: facade.deps && facade.deps.map(convertR3DependencyMetadata),
+        },
+        /* resolveForwardRefs */ false);
```
The reason why this can be `false` is very subtle here; I initially thought this was incorrect but the forward-ref unwrapping has happened during the creation of the `R3InjectableMetadataFacade` object:

https://github.com/angular/angular/blob/2d3cd2b96908f73d0c6fa3c309ac184b3ced0461/packages/core/src/di/jit/injectable.ts#L102-L114

That may deserve a comment here.

petebacondarwin replied:
Perhaps we should not resolve them in `jit/injectable`??

JoostK replied:
I don't mind either way

petebacondarwin replied:
I think it is not good to have resolving code in two places, only to disable it in the second place. I've removed it and from tidied up `jit/injectable.ts`.

JoostK commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -445,6 +468,19 @@ function parseJitTemplate(
+/**
+ * Convert the expression, if present` to an `R3ProviderExpression`.
+ *
+ * In JIT mode we do not know, or care, whether the expression was wrapped in a `forwardRef()`.
```
Well, we sort of know through `isForwardRef`; it's indeed more that we don't care _because_ `resolveForwardRefs = true` is used. That is needed to delay the unwrapping of potential forward refs until the factory is invoked instead of right away, as resolving it here would be too early.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -367,6 +367,48 @@ export enum FactoryTarget {
+   * - `InjectorType` such as `NgModule`,
+   * - `'root'` the root injector
+   * - `'any'` all injectors.
+   * - `null`, does not belong to any injector. Must be explicitly listed in the injector
```
I think that `null` is never generated because of that `if ((meta.providedIn as o.LiteralExpr).value !== null)` logic in the compiler.

JoostK commented on packages/compiler/src/render3/partial/injectable.ts:
```diff
@@ -0,0 +1,89 @@
+ * If `R3ProviderExpression.isForwardRef` is true then the expression was originally wrapped in a
+ * `forwardRef()` call to prevent the value from being eagerly evaluated in the code.
+ *
+ * Normally, the linker will statically process the code, putting the `expression` inside a factory
+ * function so the `forwardRef()` wrapper is not evaluated before it has been defined. But if the
+ * partial declaration is evaluated by the JIT compiler the `forwardRef()` call is still needed to
+ * prevent eager evaluation of the `expression`.
+ *
+ * So in partial declarations, expressions that could be forward-refs are wrapped in `forwardRef()`
+ * calls, and this is then unwrapped in the linker as necessary.
```
These are _excellent_ comments!

JoostK reviewed (commented):
This looks great!

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_injectable_linker_1.ts:
```diff
@@ -42,25 +42,25 @@ export function toR3InjectableMeta<TExpression>(
   if (metaObj.has('useClass')) {
     meta.useClass = extractForwardRef(metaObj.getValue('useClass')) ??
-        createR3ProviderExpression(metaObj.getOpaque('useClass'));
+        createR3ProviderExpression(metaObj.getOpaque('useClass'), /* isForwardRef */ false);
```
Can we make `extractForwardRef` always return a `R3ProviderExpression`, or perhaps have an additional function that always returns `createR3ProviderExpression`?

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_component_linker_1.ts:
```diff
@@ -116,13 +114,11 @@ export class PartialComponentLinkerVersion1<TStatement, TExpression> implements
-        const forwardRefType = extractForwardRef(pipe);
-        if (forwardRefType !== null) {
+        const {expression: pipeType, isForwardRef} = extractForwardRef(pipe);
+        if (isForwardRef !== null) {
```
This looks like a bug

petebacondarwin replied:
Doh!

AndrewKushnir commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -60,18 +60,41 @@ export class CompilerFacadeImpl implements CompilerFacade {
+          useFactory: wrapExpression(facade, USE_FACTORY),
+          useValue: convertToProviderExpression(facade, USE_VALUE),
+          useExisting: convertToProviderExpression(facade, USE_EXISTING),
+          deps: facade.deps && facade.deps.map(convertR3DependencyMetadata),
```
nit:
```suggestion
          deps: facade.deps?.map(convertR3DependencyMetadata),
```

petebacondarwin replied:
👍  originally `deps` could be `null` so it made sense to use `&&`. But now it can only be `undefined` so `??` is more accurate.

AndrewKushnir commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -60,18 +60,41 @@ export class CompilerFacadeImpl implements CompilerFacade {
+          useFactory: wrapExpression(facade, USE_FACTORY),
+          useValue: convertToProviderExpression(facade, USE_VALUE),
+          useExisting: convertToProviderExpression(facade, USE_EXISTING),
+          deps: facade.deps && facade.deps.map(convertR3DeclareDependencyMetadata),
```
```suggestion
          deps: facade.deps?.map(convertR3DeclareDependencyMetadata),
```

AndrewKushnir commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -446,22 +469,38 @@ function parseJitTemplate(
 
-function wrapExpression(obj: any, property: string): WrappedNodeExpr<any>|undefined {
+/**
+ * Convert the expression, if present` to an `R3ProviderExpression`.
```
```suggestion
 * Convert the expression, if present to an `R3ProviderExpression`.
```

AndrewKushnir commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -367,6 +367,48 @@ export enum FactoryTarget {
+   * If not provided, then it does not belong to any injector. Must be explicitly listed in the
+   * providers of an injector.
+   */
+  providedIn?: o.Expression;
+  /**
```
nit: may be add a new line after the field and before the next field's comment (for slightly better readability and consistency with other interfaces format)?
```suggestion
  providedIn?: o.Expression;
  
  /**
```

AndrewKushnir commented on packages/compiler/src/render3/partial/util.ts:
```diff
@@ -46,3 +50,46 @@ export function toOptionalLiteralMap<T>(
+
+export function compileDependencies(deps: R3DependencyMetadata[]|'invalid'|null): o.LiteralExpr|
+    o.LiteralArrayExpr {
+  if (deps === 'invalid') {
```
I think it might be helpful (as an additional context) to have a reference here to point to the code where the `invalid` string is produced.

petebacondarwin replied:
Good call. It even just took me 5 mins to find the line of code where this can happen!

AndrewKushnir commented on packages/core/src/di/jit/injectable.ts:
```diff
@@ -47,12 +46,11 @@ export function compileInjectable(type: Type<any>, srcMeta?: Injectable): void {
-            typeArgumentCount: metadata.typeArgumentCount,
+            name: type.name,
+            type,
+            typeArgumentCount: 0,
```
May be add a comment on why it's always `0`?

AndrewKushnir commented on packages/core/test/acceptance/providers_spec.ts:
```diff
@@ -657,8 +657,10 @@ describe('providers', () => {
-      TestBed.configureTestingModule(
-          {declarations: [App], providers: [{provide: SomeProvider, useClass: SomeProviderImpl}]});
+      // We don't configure the `SomeProvider` in the TestingModule so that it uses the
+      // tree-shakable provider given in the `@Injectable` decorator above, which makes use fo the
```
```suggestion
      // tree-shakable provider given in the `@Injectable` decorator above, which makes use of the
```

AndrewKushnir reviewed (approved):
LGTM 👍

AndrewKushnir commented:
Reviewed-for: public-api

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented:
[Presubmit](http://test/OCL:367100791:BASE:367100819:1617748994137:6840b719).

jessicajaniuk commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/util.ts:
```diff
@@ -29,3 +30,66 @@ export function parseEnum<TExpression, TEnum>(
+/**
+ * Parse a dependency structure from an AST object.
+ */
+export function getDep<TExpression>(depObj: AstObject<R3DeclareDependencyMetadata, TExpression>):
```
I know this is how it was named before, but I would prefer this to be `getDependency` rather than the abbreviation. Not that big of a deal though.

petebacondarwin replied:
Since I created the original `getDeps()` fn I shall indeed rename it.

jessicajaniuk reviewed (approved):
LGTM :cookie:

Just one loose suggestion that can be ignored.

reviewed-for: public-api

petebacondarwin committed:
fixup! refactor(compiler): implement `ngDeclareInjectable()`

alxhub reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:367100791:BASE:367235383:1617812889891:480cfbb8).

AndrewKushnir commented:
@petebacondarwin FYI, presubmits look good.
